In [2]:
USE hr_db
GO

Commands completed successfully.

Total execution time: 00:00:00.002

In [2]:
SELECT TOP 10 * FROM Employees;

(10 rows affected)

Total execution time: 00:00:00.046

EmpID Age AgeGroup Attrition BusinessTravel DailyRate Department DistanceFromHome Education EducationField EmployeeCount EmployeeNumber EnvironmentSatisfaction Gender HourlyRate JobInvolvement JobLevel JobRole JobSatisfaction MaritalStatus MonthlyIncome SalarySlab MonthlyRate NumCompaniesWorked Over18 OverTime PercentSalaryHike PerformanceRating RelationshipSatisfaction StandardHours StockOptionLevel TotalWorkingYears TrainingTimesLastYear WorkLifeBalance YearsAtCompany YearsInCurrentRole YearsSinceLastPromotion YearsWithCurrManager 297.00 18 18-25 1 Travel_Rarely 230 Research & Development 3 3 Life Sciences 1 405 3 Male 54 3 1 Laboratory Technician 3 Single 1420 Upto 5k 25233 1 Y 0 13 3 3 80 0 0 2 3 0 0 0 0 302.00 18 18-25 0 Travel_Rarely 812 Sales 10 3 Medical 1 411 4 Female 69 2 1 Sales Representative 3 Single 1200 Upto 5k 9724 1 Y 0 12 3 1 80 0 0 2 3 0 0 0 0 458.00 18 18-25 1 Travel_Frequently 1306 Sales 5 3 Marketing 1 614 2 Male 69 3 1 Sales Representative 2 Single 1878 Upto 5k 8059 1 Y 1 14 3 4 80 0 0 3 3 0 0 0 0 728.00 18 18-25 0 Non-Travel 287 Research & Development 5 2 Life Sciences 1 1012 2 Male 73 3 1 Research Scientist 4 Single 1051 Upto 5k 13493 1 Y 0 15 3 4 80 0 0 2 3 0 0 0 0 829.00 18 18-25 1 Non-Travel 247 Research & Development 8 1 Medical 1 1156 3 Male 80 3 1 Laboratory Technician 3 Single 1904 Upto 5k 13556 1 Y 0 12 3 4 80 0 0 0 3 0 0 0 0 973.00 18 18-25 0 Non-Travel 1124 Research & Development 1 3 Life Sciences 1 1368 4 Female 97 3 1 Laboratory Technician 4 Single 1611 Upto 5k 19305 1 Y 0 15 3 3 80 0 0 5 4 0 0 0 0 1154.00 18 18-25 1 Travel_Frequently 544 Sales 3 2 Medical 1 1624 2 Female 70 3 1 Sales Representative 4 Single 1569 Upto 5k 18420 1 Y 1 12 3 3 80 0 0 2 4 0 0 0 0 1312.00 18 18-25 0 Non-Travel 1431 Research & Development 14 3 Medical 1 1839 2 Female 33 3 1 Research Scientist 3 Single 1514 Upto 5k 8018 1 Y 0 16 3 3 80 0 0 4 1 0 0 0 0 128.00 19 18-25 1 Travel_Rarely 528 Sales 22 1 Marketing 1 167 4 Male 50 3 1 Sales Representative 3 Single 1675 Upto 5k 26820 1 Y 1 19 3 4 80 0 0 2 2 0 0 0 0 150.00 19 18-25 0 Travel_Rarely 1181 Research & Development 3 1 Medical 1 201 2 Female 79 3 1 Laboratory Technician 2 Single 1483 Upto 5k 16102 1 Y 0 14 3 4 80 0 1 3 3 1 0 0 0

In [5]:
SELECT COUNT(*) as Count_rows 
FROM Employees

(1 row affected)

Total execution time: 00:00:00.013

Count_rows 1480

In [4]:
SELECT EmpID, COUNT(*) AS count
FROM Employees
GROUP BY EmpID
HAVING COUNT(*) > 1;

(10 rows affected)

Total execution time: 00:00:00.027

EmpID count 1461.00 2 1462.00 2 1463.00 2 1464.00 2 1465.00 2 1466.00 2 1467.00 2 1468.00 2 1469.00 2 1470.00 2

Note: Total employees recorded in db is 1480-10 = 1470 employees

In [6]:
SELECT COUNT(*) AS count, Gender
FROM (
    SELECT DISTINCT EmpID, Gender
    FROM Employees
) AS UniqueEmployees
GROUP BY Gender;


(2 rows affected)

Total execution time: 00:00:00.910

count Gender 588 Female 882 Male

 Show department and their role

In [7]:
SELECT Department, JobRole
FROM Employees
GROUP BY Department, JobRole
ORDER BY Department;

(11 rows affected)

Total execution time: 00:00:13.516

Department JobRole Human Resources Human Resources Human Resources Manager Research & Development Healthcare Representative Research & Development Laboratory Technician Research & Development Manager Research & Development Manufacturing Director Research & Development Research Director Research & Development Research Scientist Sales Manager Sales Sales Executive Sales Sales Representative

Check some value of dataset

In [8]:
SELECT ROUND(AVG(Age),2) AS Avg_age, Department
FROM (
    SELECT DISTINCT EmpID, Age, Department
    FROM Employees
) AS UniqueEmployees
GROUP BY Department

(3 rows affected)

Total execution time: 00:00:01.591

Avg_age Department 37 Human Resources 37 Research & Development 36 Sales

In [9]:
SELECT ROUND(AVG(Yearsatcompany), 2) AS AVG_years_at_company
FROM (
    SELECT DISTINCT EmpID, Yearsatcompany
    FROM Employees
) AS UniqueEmployees

(1 row affected)

Total execution time: 00:00:00.019

AVG_years_at_company 7

In [10]:
SELECT ROUND(AVG(MonthlyIncome),2) AS AverageMonthlyRate
FROM (
    SELECT DISTINCT EmpID, MonthlyIncome
    FROM Employees
) AS UniqueEmployees

(1 row affected)

Total execution time: 00:00:00.019

AverageMonthlyRate 6502

Analyze attrition

In [14]:
SELECT COUNT(EmpID) AS Total_Attrition
FROM (
    SELECT DISTINCT EmpID, Attrition
    FROM Employees
) AS UniqueEmployees
WHERE Attrition = '1';

(1 row affected)

Total execution time: 00:00:00.836

Total_Attrition 237

238/1470 about 16% employees was leave

In [16]:
SELECT COUNT(EmpID) As Num_Employees, AgeGroup, ROUND(AVG(MonthlyIncome),2) As Avg_Income 
FROM (
    SELECT DISTINCT EmpID, AgeGroup, MonthlyIncome , Attrition
    FROM Employees
) AS UniqueEmployees
WHERE Attrition = '1'
GROUP BY AgeGroup

(5 rows affected)

Total execution time: 00:00:03.264

Num_Employees AgeGroup Avg_Income 44 18-25 2575 116 26-35 4174 43 36-45 6001 26 46-55 8637 8 55+ 6789

- The average salary of people who have left their jobs is generally lower than the median salary, so that could be one of the reasons.
- A high percentage of 25-35 year olds indicates that the company's growth opportunities are in trouble.

In [18]:
WITH MaxIncomePerAgeGroup AS (
    SELECT AgeGroup, MAX(MonthlyIncome) AS MaxSalary
    FROM (
    SELECT DISTINCT EmpID, AgeGroup, MonthlyIncome, Attrition
    FROM Employees
) AS UniqueEmployees  
    WHERE Attrition = '1'
    GROUP BY AgeGroup
)
SELECT e.EmpID, e.AgeGroup, e.MonthlyIncome AS MaxSalary
FROM (
    SELECT DISTINCT EmpID, AgeGroup, MonthlyIncome, Attrition
    FROM Employees
) AS e
JOIN MaxIncomePerAgeGroup m
ON e.AgeGroup = m.AgeGroup AND e.MonthlyIncome = m.MaxSalary
WHERE e.Attrition = '1';


(5 rows affected)

Total execution time: 00:00:00.024

EmpID AgeGroup MaxSalary 596.00 55+ 19246 569.00 46-55 19859 46.00 36-45 19545 436.00 26-35 13610 108.00 18-25 5744

In [19]:
WITH MaxIncomePerAgeGroup AS (
    SELECT AgeGroup, MIN(MonthlyIncome) AS MaxSalary
    FROM (
    SELECT DISTINCT EmpID, AgeGroup, MonthlyIncome, Attrition
    FROM Employees
) AS UniqueEmployees  
    WHERE Attrition = '1'
    GROUP BY AgeGroup
)
SELECT e.EmpID, e.AgeGroup, e.MonthlyIncome AS MaxSalary
FROM (
    SELECT DISTINCT EmpID, AgeGroup, MonthlyIncome, Attrition
    FROM Employees
) AS e
JOIN MaxIncomePerAgeGroup m
ON e.AgeGroup = m.AgeGroup AND e.MonthlyIncome = m.MaxSalary
WHERE e.Attrition = '1';


(5 rows affected)

Total execution time: 00:00:00.020

EmpID AgeGroup MaxSalary 514.00 18-25 1009 1339.00 26-35 1081 1084.00 36-45 2018 780.00 46-55 2461 1445.00 55+ 2339

Income with each role in department

In [20]:
SELECT COUNT(EmpID) As Num_Employees, Department, JobRole, ROUND(AVG(MonthlyIncome),2) As Avg_Income 
FROM (
    SELECT DISTINCT EmpID, Department, JobRole, MonthlyIncome, Attrition
    FROM Employees
) AS UniqueEmployees
WHERE Attrition = '1'
GROUP BY Department, JobRole
ORDER BY Department, JobRole;

(10 rows affected)

Total execution time: 00:00:00.013

Num_Employees Department JobRole Avg_Income 12 Human Resources Human Resources 3715 9 Research & Development Healthcare Representative 8548 62 Research & Development Laboratory Technician 2919 3 Research & Development Manager 15106 10 Research & Development Manufacturing Director 7365 2 Research & Development Research Director 19395 47 Research & Development Research Scientist 2780 2 Sales Manager 19334 57 Sales Sales Executive 7489 33 Sales Sales Representative 2364

Overtime status

In [21]:
SELECT Department, JobRole, ROUND(COUNT(CASE WHEN OverTime = '1' THEN 1 END) * 100.0 / COUNT(*),2) AS PercentOverTime
FROM (
    SELECT DISTINCT EmpID, Department, JobRole, OverTime
    FROM Employees
) AS UniqueEmployees
GROUP BY Department,JobRole
ORDER BY Department,JobRole;


Warning: Null value is eliminated by an aggregate or other SET operation.

(11 rows affected)

Total execution time: 00:00:01.540

Department JobRole PercentOverTime Human Resources Human Resources 25.000000000000 Human Resources Manager 36.360000000000 Research & Development Healthcare Representative 28.240000000000 Research & Development Laboratory Technician 23.940000000000 Research & Development Manager 24.070000000000 Research & Development Manufacturing Director 26.900000000000 Research & Development Research Director 28.750000000000 Research & Development Research Scientist 33.220000000000 Sales Manager 27.030000000000 Sales Sales Executive 28.830000000000 Sales Sales Representative 28.920000000000

In [24]:
SELECT Department, Gender, AVG(WorkLifeBalance) AS AvgWorkLifeBalance
FROM (
    SELECT DISTINCT EmpID, Gender, Department, WorkLifeBalance
    FROM Employees
) AS UniqueEmployees
GROUP BY Department, Gender
ORDER BY Department, Gender;



(6 rows affected)

Total execution time: 00:00:00.503

Department Gender AvgWorkLifeBalance Human Resources Female 2 Human Resources Male 2 Research & Development Female 2 Research & Development Male 2 Sales Female 2 Sales Male 2

Both Man or woman have a problem for balance their life and work. We need to talk with their manager and ask them something to solve this problem

Job Satisfaction status

In [25]:
SELECT e1.Department,
       e1.AvgMonthlyIncome,
       e2.AvgJobSatisfaction
FROM (
    SELECT Department, AVG(MonthlyIncome) AS AvgMonthlyIncome
    FROM (
    SELECT DISTINCT EmpID, Department, MonthlyIncome
    FROM Employees
) AS UniqueEmployees
    GROUP BY Department
) AS e1
JOIN (
    SELECT Department, AVG(JobSatisfaction) AS AvgJobSatisfaction
    FROM (
    SELECT DISTINCT EmpID, Department, JobSatisfaction
    FROM Employees
) AS UniqueEmployees
    GROUP BY Department
) AS e2 ON e1.Department = e2.Department
ORDER BY e1.AvgMonthlyIncome DESC;


(3 rows affected)

Total execution time: 00:00:00.741

Department AvgMonthlyIncome AvgJobSatisfaction Sales 6959 2 Human Resources 6654 2 Research & Development 6281 2

In [26]:
SELECT e1.JobRole,
       e1.AvgMonthlyIncome,
       e2.AvgJobSatisfaction
FROM (
    SELECT JobRole, AVG(MonthlyIncome) AS AvgMonthlyIncome
    FROM (
    SELECT DISTINCT EmpID, JobRole, MonthlyIncome
    FROM Employees
) AS UniqueEmployees
    GROUP BY JobRole
) AS e1
JOIN (
    SELECT JobRole, AVG(JobSatisfaction) AS AvgJobSatisfaction
    FROM (
    SELECT DISTINCT EmpID, JobRole, JobSatisfaction
    FROM Employees
) AS UniqueEmployees
    GROUP BY JobRole
) AS e2 ON e1.JobRole = e2.JobRole
ORDER BY e1.AvgMonthlyIncome DESC;


(9 rows affected)

Total execution time: 00:00:00.022

JobRole AvgMonthlyIncome AvgJobSatisfaction Manager 17181 2 Research Director 16033 2 Healthcare Representative 7528 2 Manufacturing Director 7295 2 Sales Executive 6924 2 Human Resources 4235 2 Research Scientist 3239 2 Laboratory Technician 3237 2 Sales Representative 2626 2

In [27]:
SELECT e.JobRole,
       e.AvgJobSatisfaction,
       o.OverTimeRate,
       e.AvgWorkLifeBalance
FROM (
    SELECT JobRole, AVG(JobSatisfaction) AS AvgJobSatisfaction, AVG(WorkLifeBalance) AS AvgWorkLifeBalance
    FROM (
    SELECT DISTINCT EmpID, JobRole, JobSatisfaction, WorkLifeBalance
    FROM Employees
) AS UniqueEmployees
    GROUP BY JobRole
) AS e
JOIN (
    SELECT JobRole, SUM(CASE WHEN OverTime = '1' THEN 1 ELSE 0 END) * 100.0 / COUNT(*) AS OverTimeRate
    FROM (
    SELECT DISTINCT EmpID, JobRole, OverTime
    FROM Employees
) AS UniqueEmployees
    GROUP BY JobRole
) AS o ON e.JobRole = o.JobRole
ORDER BY e.AvgJobSatisfaction DESC, o.OverTimeRate DESC;


(9 rows affected)

Total execution time: 00:00:00.028

JobRole AvgJobSatisfaction OverTimeRate AvgWorkLifeBalance Research Scientist 2 33.219178082191 2 Sales Representative 2 28.915662650602 2 Sales Executive 2 28.834355828220 2 Research Director 2 28.750000000000 2 Healthcare Representative 2 28.244274809160 2 Manufacturing Director 2 26.896551724137 2 Manager 2 26.470588235294 2 Human Resources 2 25.000000000000 2 Laboratory Technician 2 23.938223938223 2

We can see the correlation between the numbers above, overtime rate also highly affects employee evaluations.